[Content] Add diogogpinto-filament-themes-full-guide.md content and respective art (#587)
filamentphp/filamentphp.com: content/articles/diogogpinto-filament-themes-full-guide.md

    ],
}
```

Additionally, in the case of the `Curator` plugin, the author requires you to import some CSS files in your `theme.css`. This can be done, by editing the file `resources/css/filament/admin/theme.css` and making it look like below:

```css
@import '/vendor/filament/filament/resources/css/theme.css';
@import '/node_modules/cropperjs/dist/cropper.css';
@import '/vendor/awcodes/filament-curator/resources/css/plugin.css';

@config 'tailwind.config.js';
```

This last step imports all custom classes needed by the plugin to render correctly. It imports the class, puts into our panel's unique CSS file and minimizes it all for greater performance.

#### Running the build process

When installing plugins and editing any theme related files, always remember to run `npm run build` when you're finished to compile all the assets.

## Customizing the theme

Laravel, and Filament being Laravel based, offer multiple solutions to the same problem. So, we will address two ways of building your own Filament look and feel, the recommended way and the not-so-recommended way. Let's start with the latter.

### Before starting the customization process

> You should start the `npm run dev` process in the root of your project. This process watches for changes in the files registered in the `vite.config.js` (and `tailwind.config.js` by extension) and automatically builds a temporary CSS file to render the changes on file save.

Reviewer: Flow suggestions

```suggestion
While you develop your theme, you should run `npm run dev` in the root of your project. This process watches for changes in the files registered in the `vite.config.js` (and `tailwind.config.js`, by extension) and automatically renders any changes on file save.
```

> [!note]

> > You still need to run `npm run build` when you're done customizing your theme

Reviewer: ```suggestion
> You still need to run `npm run build` when you're finished customizing your theme
```

### The recommended way to customize your theme

> The Filament panels core offers you a list of classes. Every component of the panel layout, including forms, infolists and tables, were wrapped in custom classes to make it possible, for any developer, to customize the panel look and feel. Let's look at the following example:

Reviewer: ```suggestion
Every component in the Panel Builder package, including the layout, forms, infolists and tables, are wrapped in custom classes to make it possible to customize the panel's look and feel. Let's look at the following example:
```

![Breadcrumb Example](/images/content/articles/diogogpinto-filament-themes-full-guide/breadcrumb-example.webp)

> Looking at the breadcrumb component with Chrome's builtin Developer Tools > Inspect Element, you can see that the `<nav>`na tag contains the following classes:

Reviewer: ```suggestion
Looking at the breadcrumb component with Chrome's built-in Developer Tools > Inspect Element, you can see that the `<nav>` tag contains the following classes:
```

```html
<nav class="fi-breadcrumbs mb-2 hidden sm:block">
```

> The `.fi-breadcrumbs` class is inserted to allow developers to further optimize the breadcrumbs' look and feel. By default, the class has no CSS properties defined in itself, it's only there for the developer to apply custom styles.

Reviewer: ```suggestion
The `.fi-breadcrumbs` class allows us to further optimize the breadcrumbs' look and feel. By default, the class has no CSS properties defined–it only exists for us to apply custom styles.
```

#### Customizing the CSS

> Picking up in the above example of `.fi-breadcrumbs`, we can now apply custom CSS classes on our `theme.css`. You can either use Tailwind utility classes or just write vanilla CSS. Let's say you want to override the default margin `mb-2`. Inputting the below CSS will yield the exact same results:

Reviewer: ```suggestion
Using the above example of `.fi-breadcrumbs`, we now apply custom CSS classes in our `theme.css` file to change the styling of the breadcrumbs. You can either use Tailwind utility classes or vanilla CSS. Let's say you want to override the default margin `mb-2`. Either option in the below CSS will yield the exact same results:
```

```css
.fi-breadcrumbs {
    @apply mb-4;
}

.fi-breadcrumbs {
    margin-bottom: 1rem;

> }

Reviewer: I actually think this isn't _necessarily_ always going to work.

I believe that, specifically referencing the bottom margin, we won't reliably be able to override the CSS styling. This is because there is already a `mb-2` class in the class string, so when our styles get applied via `.fi-breadcrumbs`, `mb-2` and our `mb-4` will (theoretically) have the same specificity. If _that_ is the case, whichever class is defined further down in the cascade with take precedence.

I might be wrong here, but regardless it may be worth changing this example from an "overriding" example to an "adding" example. Something like adding a top padding, border, etc.

@zepfietje (or anyone from Core), let me know if I'm completely incorrect here.

Author: @alexandersix just wondering, given that the custom classes are always defined after TW classes, wouldn't they always take precedence? Or is there a case where that may not be true?

I may be missing something

Reviewer 2: If you’re overriding an existing style, the user will need to use `!important` or the `!` prefix in the Tailwind classes.

For example, if we’re changing the sidebar color, this code won’t work:

```css
.fi-sidebar {
  @apply bg-white;
}
```
To ensure it takes precedence over Filament’s predefined styles, we need to mark the class as important:

```css
.fi-sidebar {
  @apply !bg-white;
}
```

```

> As the `.fi-breadcrumbs` class is inserted before the default Tailwind utility classes, applying custom CSS will always override the default style.

Reviewer: See above comment, but I believe the order in which a class string is laid out has no bearing on specificity or the cascade, so having multiple classes _of the same specificity_ may lead to unreliable outputs based on where the Tailwind classes are defined in relation to one another.

Reviewer 3: Yup, totally depends on specificity.

Author: @zepfietje @alexandersix 

I'm probably missing something, given that both TW classes and custom classes created with @apply are technically single class selectors, shouldn’t they all have equal specificity?

I clearly made a mistake in the text, as it isn't order of declaration that matters, but the build process that builds custom classes after the utility classes.

Hope to get some clarity, thank you in advance!

Author: Do you think this makes sense?

#### Diving deeper

If we dive deeper into the breadcrumbs, we will find all the following classes, in this hierarchy:

```
- .fi-breadcrumbs
-- .fi-breadcrumbs-list
--- .fi-breadcrumbs-item
---- .fi-breadcrumbs-item-separator
---- .fi-breadcrumbs-item-label
```

Let's say we want to really change the breadcrumbs appearance. We can style each of this classes. Check the following example:

```css
.fi-breadcrumbs-list {
    @apply gap-x-0;
}

.fi-breadcrumbs-item {
    @apply gap-x-0;
}

.fi-breadcrumbs-item-label {
    @apply font-thin dark:text-primary-500 text-primary-500;
}

span.fi-breadcrumbs-item-label {
    @apply border-b-2 border-b-primary-500 font-semibold -mb-[2px];

> }

Reviewer: It is confusing as to why you have the same class twice, with one being more specific due to the `span` selector. Could you combine these two together?

Author: @alexandersix I believe in this case breadcrumbs active item is wrapped in a `<span>` tag, the others on a `<a>` tag.

Maybe something like the following would be better?

```css
.fi-breadcrumbs-list > li:last-child .fi-breadcrumbs-item-label
```

Will have to try it out.

.fi-breadcrumbs-item-separator {
    @apply fill-primary-500 h-4;
}
```

![Breadcrumbs Final Example](/images/content/articles/diogogpinto-filament-themes-full-guide/breadcrumbs-final.webp)

> There are a lot of classes around every Filament component. You can go through each one and apply your custom CSS, so you can create a unique design for your Panel.

Reviewer: ```suggestion
There are many classes within each Filament component. You can go through each one and apply your custom CSS to create a unique design for your Panel.
```

> [!TIP]

> > There's a useful Filament Plugin ([Theme Inspector](https://filamentphp.com/plugins/codewithdennis-theme-inspector)) that lets you inspect all the classes in your panel in your dev environment. You can even copy to clipboard! It really smoothes the development process.

Reviewer: ```suggestion
> There's a useful Filament Plugin ([Theme Inspector](https://filamentphp.com/plugins/codewithdennis-theme-inspector)) that lets you inspect all the classes in your panel in your dev environment. You can even copy to clipboard! It really smoothes out the theme development process.
```

> ### Going behind CSS

Reviewer: Is this maybe a better word for what you're trying to get across?

```suggestion
### Going beyond CSS
```

> Sometimes there are some things your theme requires that go behind CSS. You may want to add a custom Livewire component, Blade View or JS script. As these modifications are outside the scope of this guide, I would advise you to check the following resources:

Reviewer: ```suggestion
Sometimes there are some things your theme requires that go beyond CSS. You may want to add a custom Livewire component, Blade View, or JS script. All of this is possible, but as these modifications are outside the scope of this guide, I would advise you to check the following resources:
```

> - [Render Hooks on the Official Docs](https://filamentphp.com/docs/3.x/support/render-hooks) - check how you can render custom views almost anywhere in your panels

Reviewer: ```suggestion
- [Render Hooks on the Official Docs](https://filamentphp.com/docs/3.x/support/render-hooks) - Render custom views almost anywhere in your panels
```

> - [Registering Assets](https://filamentphp.com/docs/3.x/support/assets#registering-javascript-files) - check how you can register custom JS/CSS into your Filament panels

Reviewer: ```suggestion
- [Registering Assets](https://filamentphp.com/docs/3.x/support/assets#registering-javascript-files) - Register custom CSS/JS in your Filament panels
```

> ### The ~wrong~ unrecommended way to customize your theme

Reviewer: ```suggestion
### The non-recommended way to customize your theme
```

> ### The ~wrong~ unrecommended way to customize your theme

Reviewer 3: We really should not have this section in my opinion as this approach should basically never be used.

Reviewer: I'm on the fence about this one. On the one hand, I completely agree that we almost never want this approach to be used. However, on the other hand, by having this here, we're specifically calling out the fact that this is the _wrong_ approach, dissuading people from using it.

I guess it comes down to whether or not this is going to prompt more people to remember that this is an option and use it or if it's going to prevent people from using it because it's explicitly stated not to.

I think I come down more on the side of leave it in to ensure that people know it's the wrong way to do things. My hunch is that the people reading this article are (in general) going to be deep enough in Filament to know that overriding views is an option in Laravel 🤷🏻

Author: I can see both points!

I inserted because sometimes you may want to change specific functionality around whole components that may not necessarily break functionality on updates. My case was the topbar, but later I disabled it and used render hooks. But that approach gave me some headaches around mobile and had to create a whole new mobile trigger for the sidebar - if I had to do it all over again, I would publish the file and monitor each release for breaking changes.

Another way of doing things, is publishing all the views of the Filament panels package. Why isn't this recommended? The docs say exactly why:

> You may be tempted to publish the internal Blade views to your application so that you can customize them. We don't recommend this, as it will introduce breaking changes into your application in future updates. Please use the CSS hook classes wherever possible.
> If you do decide to publish the Blade views, please lock all Filament packages to a specific version in your composer.json file, and then update Filament manually by bumping this number, testing your entire application after each update. This will help you identify breaking changes safely.

> [!warning]
> This is for demonstration purposes only, you really shouldn't do this as it can break your panels in future releases/updates. If you need to do this, only publish the necessary files and lock filament to your project's current version

So, if you were to do this, you could run the following command in your terminal:

```bash
php artisan vendor:publish --tag=filament-panels-views
```

> Now, inside your `resources/views/vendor/filament-panels` folder are a bunch of files - these are the original view files you can (but shouldn't) customize.

Reviewer: ```suggestion
Now, you will find a bunch of files inside your `resources/views/vendor/filament-panels` folder. These are the original view files you can (but shouldn't) customize.
```

![Vendor Files](/images/content/articles/diogogpinto-filament-themes-full-guide/vendor-publish.webp)

Navigate to your themes `tailwind.config.js` and add the filament-panels folder to the content array, so vite can process these file changes. If you're following along, your file should look like this:

```js
import preset from '../../../../vendor/filament/filament/tailwind.config.preset'

export default {
    presets: [preset],
    content: [
        './app/Filament/Clusters/Products/**/*.php',
        './resources/views/filament/clusters/products/**/*.blade.php',
        './vendor/filament/**/*.blade.php',
        './vendor/diogogpinto/filament-auth-ui-enhancer/resources/**/*.blade.php',
        './vendor/awcodes/filament-curator/resources/**/*.blade.php',

>         // Added the following line of code
>         './resources/views/filament-panels/**/*.blade.php',

Reviewer: ```suggestion
        './resources/views/filament-panels/**/*.blade.php', // [tl! add]
```

    ],
}
```

#### Example - make the topbar backdrop blur

If you're editing the blade files directly, to blur the background of a topbar, you should open the following file:

```
./resources/views/vendor/filament-panels/components/topbar/index.php
```

Now we can edit the `<nav>` tag, and make it look like the following:

```html
<nav
    class="flex h-16 items-center gap-x-4 bg-white bg-opacity-50 backdrop-blur-lg px-4 shadow-sm ring-1 ring-gray-950/5 dark:bg-gray-900 dark:bg-opacity-50 dark:ring-white/10 md:px-6 lg:px-8"
>
```

To make it look consistent in the sidebar, we should also open the following file:

```
./resources/views/vendor/filament-panels/components/sidebar/index.php
```

And change the `<header>` tag to make it look like:

```html
<header
    class="fi-sidebar-header flex h-16 items-center bg-white bg-opacity-50 backdrop-blur-lg px-6 ring-1 ring-gray-950/5 dark:bg-gray-900 dark:bg-opacity-50 dark:ring-white/10 lg:shadow-sm"
>
```

> And there it is. We were able to customize the theme file in the unrecommended way. We shouldn't really do this, unless we want to add complex logic to the code. If this is the case, you can delete all the files except the ones you are directly going to modify.

Reviewer: ```suggestion
And there it is. We were able to customize the theme file in the unrecommended way. We shouldn't really do this, unless we want to add complex logic to our code and take on the burden of updating these files every single time Filament updates. If this is the case, you can delete all the files except the ones you are directly going to modify.
```

![Blade editing adding backdrop](/images/content/articles/diogogpinto-filament-themes-full-guide/blade-backdrop-blur.webp)

> [!note]
> All the changes you make to these files will render on ALL of your Filament panels in the project, unless you use some kind of conditional statements in your blade files.

> Remember to the things the right way whenever possible.

Reviewer: ```suggestion
Remember to do the things the right way whenever possible.
```
